node/coinstacks/solana/api/src/websocket.ts

```diff
@@ -0,0 +1,167 @@
+
+    this.pingInterval = opts?.pingInterval ?? 10000
+    this.retryAttempts = opts?.retryAttempts ?? MAX_RETRY_ATTEMPTS
+
+    this.initialize()
+  }
+
+  private initialize(): void {
+    this.socket.on('ping', () => this.socket.pong())
+    this.socket.on('pong', () => this.heartbeat())
+    this.socket.onerror = (error) => {
+      this.logger.error({ error, fn: 'ws.onerror' }, 'websocket error')
+    }
+    this.socket.onclose = ({ code, reason }) => {
+      this.logger.error({ code, reason, fn: 'ws.close' }, 'websocket closed')
+      this.close()
+    }
+    this.socket.onopen = () => this.onOpen()
+    this.socket.onmessage = (msg) => this.onMessage(msg)
+  }
+
+  private close(): void {
+    this.interval && clearInterval(this.interval)
+
+    if (++this.retryCount >= this.retryAttempts && this.retryAttempts !== 0) {
+      throw new Error('failed to reconnect')
+    }
+
+    setTimeout(
+      () => {
+        this.socket = new WebSocket(this.url, { handshakeTimeout: 5000 })
+        this.initialize()
+      },
+      Math.min(Math.random() * (BASE_DELAY * this.retryCount ** 2), MAX_DELAY)
+    )
+  }
+
+  private heartbeat(): void {
+    this.pingTimeout && clearTimeout(this.pingTimeout)
+    this.pingTimeout = setTimeout(() => {
+      this.logger.debug({ fn: 'pingTimeout' }, 'heartbeat failed')
+      this.socket.terminate()
+    }, this.pingInterval + 1000)
+  }
+
+  onOpen(): void {
+    this.logger.debug({ fn: 'ws.onopen' }, 'websocket opened')
+    this.retryCount = 0
+    this.interval = setInterval(() => this.socket.ping(), this.pingInterval)
+    this.heartbeat()
+
+    if (this.addresses.length > 0) {
+      const subscribeAddresses: Subscription = {
+        jsonrpc: '2.0',
+        id: 'newTx',
+        method: 'transactionSubscribe',
+        params: [
+          {
+            accountInclude: this.addresses,
+          },
+          {
+            encoding: 'jsonParsed',
+            transactionDetails: 'full',
+            showRewards: true,
+            maxSupportedTransactionVersion: 0,
+          },
+        ],
+      }
+      this.socket.send(JSON.stringify(subscribeAddresses))
+    }
+  }
+
+  async onMessage(message: WebSocket.MessageEvent): Promise<void> {
+    try {
+      const res: GeyserWebsocketResponse = JSON.parse(message.data?.toString() ?? message.toString())
+
+      switch (res.method) {
+        case 'transactionNotification':
+          if ('result' in res.params) {
+            if (Array.isArray(this.handleTransaction)) {
+              this.handleTransaction.map(async (handleTransaction) => handleTransaction(res.params.result.transaction))
+            } else {
+              this.handleTransaction(res.params.result.transaction)
+            }
+          }
+          return
+      }
+    } catch (err) {
+      this.logger.error(err, `failed to handle message: ${JSON.stringify(message)}`)
+    }
+  }
+
+  subscribeAddresses(addresses: string[]): void {
```

Comment:
we will want to verify this updates the subscription in place with the new set of addresses. otherwise, we will need to unsubscribe/resubscribe with the new set.

Reply:
I verified, it doesn't update the subscription in place with the new set of addresses, we need to unsubscribe before subscribing again

Very dumb but I played with it like

```js
const server = new ws('ws://api.solana.localhost')

server.on('open', function open() {
  // server.send(JSON.stringify({ subscriptionId: 'test', method: 'subscribe', data: {topic: 'txs', addresses: ['CebN5WGQ4jvEPvsVU4EoHEpgzq1VV7AbicfhtW4xC9iM']} }));
})

server.on('message', (data) => {
  console.log({time: Date.now(), message: JSON.parse(Buffer.from(data).toString())})
})

server.on('error', (data) => {
  console.log({error: data})
})


setTimeout(() => {
  console.log('subscribing')
  server.send(JSON.stringify({ subscriptionId: 'test', method: 'subscribe', data: {topic: 'txs', addresses: ['CebN5WGQ4jvEPvsVU4EoHEpgzq1VV7AbicfhtW4xC9iM']} }));
}, 2000)

setTimeout(() => {
  console.log('unsubscribing')
  server.send(JSON.stringify({ subscriptionId: 'test', method: 'unsubscribe', data: {topic: 'txs', addresses: []} }));
}, 5000)
```

Reply:
When you're talking about unsubscribe, are you talking about doing this on client side?

Because, I didn't find any unsubscribe method from the geyser websocket documentation

Reply:
Ideally they would have the ability to unsubscribe at minimum, or in place updates desired. Without this we will need to manage multiple connections unfortunately. Current connection for current address set, and upon updates to the addresses, we will need to close the current connection and open a new one with the new address set. We should reach out to their discord about this though. They are in beta for geyser and this would be something they should add and unsubscribe would be pretty easy to implement at least.

Reply:
Storing all the subIds and looping over them to unsubscribe before subscribing again in [869d1ed](https://github.com/shapeshift/unchained/pull/1053/commits/869d1ed1be75db922dd4c2150f07cec9c27f6443)